torchvision/datasets/_optical_flow.py

```diff
@@ -0,0 +1,149 @@
+import os
+from abc import ABC, abstractmethod
+from glob import glob
+from pathlib import Path
+
+import numpy as np
+import torch
+from PIL import Image
+
+from ..io.image import _read_png_16
+from .vision import VisionDataset
+
+
+__all__ = (
+    "KittiFlow",
+    "Sintel",
+)
+
+
+class FlowDataset(ABC, VisionDataset):
+    def __init__(self, root, transforms=None):
+
+        super().__init__(root=root)
+        self.transforms = transforms
+
+        self._flow_list = []
+        self._image_list = []
+
+    def _read_img(self, file_name):
+        return Image.open(file_name)
+
+    @abstractmethod
+    def _read_flow(self, file_name):
+        # Return the flow or a tuple (flow, valid) for datasets where the valid mask is built-in
+        pass
+
+    def __getitem__(self, index):
+        # Some datasets like Kitti have a built-in valid mask, indicating which flow values are valid
+        # For those we return (img1, img2, flow, valid), and for the rest we return (img1, img2, flow),
+        # and it's up to whatever consumes the dataset to decide what `valid` should be.
+
+        img1 = self._read_img(self._image_list[index][0])
+        img2 = self._read_img(self._image_list[index][1])
+        flow = self._read_flow(self._flow_list[index]) if self._flow_list else None
+
+        if isinstance(flow, tuple):
+            flow, valid = flow
+        else:
+            valid = None
+
+        if self.transforms is not None:
+            img1, img2, flow, valid = self.transforms(img1, img2, flow, valid)
```

Comment:
We're forcing all transforms to accept a `flow` and a `valid` parameter, even if those aren't used. Typically for Sintel where `valid` is always None, `valid` would always be a passthough. An example of such transform would be e.g.

```py
class RandomHorizontalFlip(T.RandomHorizontalFlip):
    def forward(self, img1, img2, flow, valid):
        if torch.rand(1) > self.p:
            return img1, img2, flow, valid

        img1 = F.hflip(img1)
        img2 = F.hflip(img2)
        flow = F.hflip(flow) * torch.tensor([-1, 1])[:, None, None]
        if valid is not None:
            valid = F.hflip(valid)
        return img1, img2, flow, valid
```

You can find more at https://github.com/NicolasHug/vision/blob/raft/references/raft/transforms.py

Reply:
Concerning `valid` being a passthrough, this is already happening in other cases so I don't think it's a problem. See:
https://github.com/pytorch/vision/blob/d367a01a18a3ae6bee13d8be3b63fd6a581ea46f/references/detection/transforms.py#L48-L54

Concerning the `self.transforms()` call, it assumes a specific API on the provided transforms. Might be worth checking with @pmeier and @fmassa that this is compatible with the new transforms RFC.

Reply:
The new API will transform everything it knows to transform. So for example if you wrap `flow` in a (not-yet-existing) `features.Flow` the transforms will ignore it unless they know how to deal with it.

Reply:
This is fine, and the new API can handle the `flow` + `valid` internally (the `valid` could be an attribute of `flow` as well)